### PR TITLE
issue: Component scan two times for @Service

### PR DESCRIPTION
**Pull request by zhangjiezhang**
Add cache to solve this issue

### CODE REVIEW

**Comment by kylixs** [comment]
Repeated scan should be a problem in the implementation logic. There are two ways to register a scanner, one is to annotate `@DubboComponentScan`/`@EnableDubbo`,  and the other is to define the configuration: `dubbo.scan.base-packages` .
A better solution is to register a single scanner (prefer to `ServiceClassPostProcessor`), then collect all the package names to be scanned from the annotations and configuration, filter the duplicate packages and sub-packages, and then scan.

**Comment by chickenlj** [comment]
Please check @kylixs 's comments.

**Comment by zhangjiezhang** [comment]
> Please check @kylixs 's comments.

My project uses ServiceClassPostProcessor and only scans once for Component's(DubboService, org.apache.dubbo.config.annotation.Service, com.alibaba.dubbo.config.annotation.Service).

org.apache.dubbo.config.spring.beans.factory.annotation.ServiceClassPostProcessor is used to run org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.findCandidateComponents(String basePackage) twice:
1.scanner.scan(packageToScan) -> scanner.doScan(basePackages) -> scanner.findCandidateComponents(basePackage): add DubboService lost by AnnotationConfigApplicationContext scan to spring container.
2.ServiceClassPostProcessor.findServiceBeanDefinitionHolders(scanner, packageToScan, registry, beanNameGenerator) -> scanner.findCandidateComponents(packageToScan): get All DubboService.

**Comment by kylixs** [comment]
Duplicated scanning is resolved in Dubbo 3 : https://github.com/apache/dubbo/blob/bdb2d318386f86de7df54c7e496fe4148691c2df/dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ServiceAnnotationPostProcessor.java#L162

**Comment by zhangjiezhang** [comment]
> Duplicated scanning is resolved in Dubbo 3 :
A0AF1E21-F218-4621-B1FA-768EF018048F
Still scan twice!!!
The scan results of the first step should be cached for the second step of query。

**Comment by kylixs** [comment]
First, avoid scan duplicated packages or sub-package:
```
  // avoid duplicated scans
  if (servicePackagesHolder.isPackageScanned(packageToScan)) {
      if (logger.isInfoEnabled()) {
          logger.info("Ignore package who has already bean scanned: " + packageToScan);
      }
      continue;
  }
```

Then, exclude scanned classes before generate BeanDefinitions:
```
ScanExcludeFilter scanExcludeFilter = new ScanExcludeFilter();        
scanner.addExcludeFilter(scanExcludeFilter);
```
```
private class ScanExcludeFilter implements TypeFilter {

    private int excludedCount;

    @Override
    public boolean match(MetadataReader metadataReader, MetadataReaderFactory metadataReaderFactory) throws IOException {
        String className = metadataReader.getClassMetadata().getClassName();
        boolean excluded = servicePackagesHolder.isClassScanned(className);
        if (excluded) {
            excludedCount ++;
        }
        return excluded;
    }

    public int getExcludedCount() {
        return excludedCount;
    }
}
```

**Comment by zhangjiezhang** [comment]
> Then, exclude scanned classes before generate BeanDefinitions:

98DAAB80-2CB8-46D7-AE93-8B7EDE9AEB50
1.the first scan: scanner.scan(packageToScan)
2.the second scan: findServiceBeanDefinitionHolders(scanner, packageToScan, registry, beanNameGenerator)
You just add classfile to servicePackagesHolder after the second scan, and you cannot avoid called  scanner.findCandidateComponents(String basePackage) twice for same class.
Because the classfile scanned for the first time has not been added to servicePackagesHolder.

**Comment by kylixs** [comment]
> 1.the first scan: scanner.scan(packageToScan)
> 2.the second scan: findServiceBeanDefinitionHolders(scanner, packageToScan, registry, beanNameGenerator)

Oh, you're right, I miss it. 

The first scan is to register the bean definition of the service implementation class, and the second scan is to register the bean definition of the dubbo ServiceBean later. So, I think the second scan is unnecessary, just use the results of the first scan directly.

Some thing like this:

```java
private void scanServiceBeans(Set<String> packagesToScan, BeanDefinitionRegistry registry) {
   ...
    for (String packageToScan : packagesToScan) {

        // avoid duplicated scans
        if (servicePackagesHolder.isPackageScanned(packageToScan)) {
            if (logger.isInfoEnabled()) {
                logger.info("Ignore package who has already bean scanned: " + packageToScan);
            }
            continue;
        }

        // Scan and registers as Spring @Service Bean first
        Set<BeanDefinitionHolder> beanDefinitionHolders = scanner.doScan(packageToScan);

        // Second, register bean definition of dubbo ServiceBean for each service
        if (!CollectionUtils.isEmpty(beanDefinitionHolders)) {
            if (logger.isInfoEnabled()) {
                List<String> serviceClasses = new ArrayList<>(beanDefinitionHolders.size());
                for (BeanDefinitionHolder beanDefinitionHolder : beanDefinitionHolders) {
                    serviceClasses.add(beanDefinitionHolder.getBeanDefinition().getBeanClassName());
                }
                logger.info("Found " + beanDefinitionHolders.size() + " classes annotated by Dubbo @Service under package [" + packageToScan + "]: " + serviceClasses);
            }
            for (BeanDefinitionHolder beanDefinitionHolder : beanDefinitionHolders) {
                processScannedBeanDefinition(beanDefinitionHolder, registry, scanner);
                servicePackagesHolder.addScannedClass(beanDefinitionHolder.getBeanDefinition().getBeanClassName());
            }
        } else {
            if (logger.isWarnEnabled()) {
                logger.warn("No class annotated by Dubbo @Service was found under package ["
                        + packageToScan + "], ignore re-scanned classes: " + scanExcludeFilter.getExcludedCount());
            }
        }
        servicePackagesHolder.addScannedPackage(packageToScan);
    }
}
```

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java [hunk 1]
```diff
@@ -67,4 +74,35 @@ public boolean checkCandidate(String beanName, BeanDefinition beanDefinition) th
         return super.checkCandidate(beanName, beanDefinition);
     }
 
+    @Override
+    public Set<BeanDefinition> findCandidateComponents(String basePackage) {
+        Set<BeanDefinition> beanDefinitions = beanDefinitionMap.get(basePackage);
```
Comment (kylixs) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java:79]: No need to cache BeanDefinition, use the result of the first scan directly outside

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java [hunk 2]
```diff
@@ -67,4 +74,35 @@ public boolean checkCandidate(String beanName, BeanDefinition beanDefinition) th
...
+
+    /**
+     * Registers @Service Bean and Finds all BeanDefinitionHolders of @Service
+     *
+     * @param registry              BeanDefinitionRegistry
+     * @param beanNameGenerator     BeanNameGenerator
+     * @param basePackage           basePackage
+     * @return                      all BeanDefinitionHolders of @Service
+     */
+    public Set<BeanDefinitionHolder> doScan(BeanDefinitionRegistry registry, BeanNameGenerator beanNameGenerator, String basePackage) {
```
Comment (kylixs) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java:95]: Do not create new `doScan` method, there is already expose a `doScan` method abrove:
```java
    @Override
    public Set<BeanDefinitionHolder> doScan(String... basePackages) {
        return super.doScan(basePackages);
    }
```

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java [hunk 3]
```diff
@@ -67,4 +74,35 @@ public boolean checkCandidate(String beanName, BeanDefinition beanDefinition) th
...
+        // Registers @Service Bean first
+        super.doScan(basePackage);
+        Set<BeanDefinition> beanDefinitions = findCandidateComponents(basePackage);
```
Comment (kylixs) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java:98]: `super.doScan(basePackage)`: will do first scanning 
`findCandidateComponents()` : will do second scanning
This did not reduce scanning.

Comment (zhangjiezhang) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java:98]: Please look i'm pull request.
Rewrite  method(Set<BeanDefinition> findCandidateComponents(String basePackage)) in DubboClassPathBeanDefinitionScanner.

Comment (zhangjiezhang) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java:98]: I rewrite method **findCandidateComponents** for DubboClassPathBeanDefinitionScanner.

**Review by zhangjiezhang — COMMENTED** [review]
Solve the cause of the problem

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java (hunk 3 above)
Comment (kylixs) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/annotation/DubboClassPathBeanDefinitionScanner.java:98]: I think just cache `Set<BeanDefinitionHolder>` in  `findCandidateComponents` method is enough,  do not need add new `doScan` method

**Comment by kylixs** [comment]
Another way is to cache the beanName and BeanDefinition in the checkCandidate method, so that there is no need to regenerate the beanName, which is more reliable.

```java
public class DubboClassPathBeanDefinitionScanner extends ClassPathBeanDefinitionScanner {

    private Set<BeanDefinitionHolder> beanDefinitionHolders = new LinkedHashSet<>();
    ....

    @Override
    public boolean checkCandidate(String beanName, BeanDefinition beanDefinition) throws IllegalStateException {
        beanDefinitionHolders.add(new BeanDefinitionHolder(beanDefinition, beanName));
        return super.checkCandidate(beanName, beanDefinition);
    }

    public Set<BeanDefinitionHolder> getBeanDefinitionHolders() {
        return beanDefinitionHolders;
    }
}

```
// ServiceAnnotationPostProcessor
```java
    private void scanServiceBeans(Set<String> packagesToScan, BeanDefinitionRegistry registry) {
      ...
        for (String packageToScan : packagesToScan) {
            ...
            // Scan and registers as Spring @Service Bean first
            // if the service class is already scanned and registered by @ComponentScan, the bean definitions will be ignored.
            scanner.scan(packageToScan);

            Set<BeanDefinitionHolder> beanDefinitionHolders = scanner.getBeanDefinitionHolders();

            // Second, register bean definition of dubbo ServiceBean for each service
            if (!CollectionUtils.isEmpty(beanDefinitionHolders)) {
            ....
      ....
}
```

**Comment by zhangjiezhang** [comment]
> Another way is to cache the `beanName` and `BeanDefinition` in the `postProcessBeanDefinition` method, so that there is no need to regenerate the beanName, which is more reliable.
> 
> ```java
> public class DubboClassPathBeanDefinitionScanner extends ClassPathBeanDefinitionScanner {
> 
>     private Set<BeanDefinitionHolder> beanDefinitionHolders = new LinkedHashSet<>();
>     ....
> 
>     @Override
>     protected void postProcessBeanDefinition(AbstractBeanDefinition beanDefinition, String beanName) {
>         beanDefinitionHolders.add(new BeanDefinitionHolder(beanDefinition, beanName));
>         super.postProcessBeanDefinition(beanDefinition, beanName);
>     }
> 
>     public Set<BeanDefinitionHolder> getBeanDefinitionHolders() {
>         return beanDefinitionHolders;
>     }
> }
> ```
> 
> // ServiceAnnotationPostProcessor
> 
> ```java
>     private void scanServiceBeans(Set<String> packagesToScan, BeanDefinitionRegistry registry) {
>       ...
>         for (String packageToScan : packagesToScan) {
>             ...
>             // Scan and registers as Spring @Service Bean first
>             // if the service class is already scanned and registered by @ComponentScan, the bean definitions will be ignored.
>             scanner.scan(packageToScan);
> 
>             Set<BeanDefinitionHolder> beanDefinitionHolders = scanner.getBeanDefinitionHolders();
> 
>             // Second, register bean definition of dubbo ServiceBean for each service
>             if (!CollectionUtils.isEmpty(beanDefinitionHolders)) {
>             ....
>       ....
> }
> ```

Yes,  you're right.👍
Either way is OK.

**Comment by kylixs** [comment]
@zhangjiezhang Please improve it and resubmit pr.

**Comment by zhangjiezhang** [comment]
> @zhangjiezhang Please improve it and resubmit pr.

Which branch to modify?

**Comment by kylixs** [comment]
Create two pull requests separately based on master and 3.0 branches, please use concise code whenever possible.

**Pull request closed without merging**